Define a platform hook interface and cleanup misuse of private calls

nashif opened the pull request:
An attempt to cleanup misuse of private API call to define platform hooks and generalize the definition of hooks implemented by platforms overriding architecture implementation.

This also generalizes some hook implementations done for some architectures and made such implementation arch agnostic, allowing hooks to be implemented by any architecture.

Relates to https://github.com/zephyrproject-rtos/zephyr/issues/58007

nashif committed:
xtensa: cache: z_soc_cached_ptr -> arch_xtensa_cached_ptr

Move from private to internal call. This should probably be promoted to
an API and implementation abstracted similar to what we do with other
cache functions.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
xtensa: irq: use BIT

use BIT macro where possible.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
riscv: sys_io: move custom sys_io to be generic

84b86d9b0 introduced the ability of defining custom sys_io functions for
riscv. This is however useful for all other architectures and there is
nothing riscv specific about it. So instead of having this be riscv
specific, move implementation to the common architecture code and make
it available for everyone using PLATFORM_HAS_CUSTOM_SYS_IO.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sys_io: rename z_soc_sys_ -> platform_sysio_

Move from private hook names to internal using a common prefix for
platform hooks.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: interrupts: rename z_soc_* to platform_*

Move from private hook names to internal using a common prefix for
platform hooks.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: riscv: Move interrupt platform hooks to use platform_ prefix.

Use platform_ instead of z_soc_ and make the feature generic supporting
all architectures with CONFIG_PLATFORM_HAS_CUSTOM_IRQ_LOCK_OPS.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: arm: use platform_ for platform hooks

Rename platform hooks enabled with Kconfig from z_soc_ to platform_.

Move from private hook names to internal using a common prefix for
platform hooks.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
xtensa: remove unused z_soc_irq_init

This call is not being used anywhere.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: document custom platform interfaces

Define a set of platform hooks.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

gmarull reviewed (commented):
changes make sense to me. we should document reserved internal prefixes like arch, platform, etc. Some CI checks would also be nice to prevent misuses (e.g. forbid `arch_`/`platform_` calls in samples/ or drivers/)

galak reviewed (commented):
any reason this can't be done with weak functions?

nashif commented:
> any reason this can't be done with weak functions?

This has been the approch for some interfaces. This PR is mostly changing the naming and moving a fewthings upo to the architecture level, but not changing how the interfaces are implemented.
We had a few change proposals floating around removing usage of _weak given it is considered not stanard but we still have many usages in the tree. Whether we use _weak or some other mechanism is something we need decide on separately I guess.

mbolivar-ampere commented on doc/hardware/porting/custom.rst:
```diff
@@ -0,0 +1,16 @@
+.. _platform_cusomizations:
+
+Hardware with Custom Interfaces
```
Seems like a good first step towards closing the documentation gap between "board porting" and "soc" or "arch" porting.

nashif committed:
manifest: update sof sha

Update with some fixes to API usage.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: rename ARM_CUSTOM_INTERRUPT_CONTROLLER and make global

Although this option is now being used only by ARM, it is needed by
other architectures, so make it global and change all occurances.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: kconfig move platform overrides together

group all custom interface configs into one menu.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: standarize *_irq_priority_set into arch_irq_priority_set

Rename all z_/z_arm_ call for setting irq priority to a new architecture
interface shared by all architectures.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: irq: define more arch_irq_ routines

define additional routines that can be shared by other architectures.
Right now this is only for ARM and will be moved up in the future.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: reorg arch interface documentation

Reorganize functions in header file for better fit.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: arm: do not call arm_gic intc directly, abstract with arch_irq

To allow other interrupt controllers to be called.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

amr-sc commented:
> > > any reason this can't be done with weak functions?
> > 
> > 
> > This has been the approch for some interfaces. This PR is mostly changing the naming and moving a fewthings upo to the architecture level, but not changing how the interfaces are implemented. We had a few change proposals floating around removing usage of _weak given it is considered not stanard but we still have many usages in the tree. Whether we use _weak or some other mechanism is something we need decide on separately I guess.
> 
> Probably more useful than having weak functions would be to fail at compile time if a `CUSTOM` option has been selected and the prescribed interface has not been implemented.
> 
> Is there a particular use for weak functions in this case?
> 
> The other bit of input I had here was that there are systems where there are multiple (system) interrupt controllers.

Hi @cfriedt , I think the _weak functions could be useful to override the default arch_ level implementations (which follow the Spec) in case some vendors doesn't conform to the Spec, or the Spec is ambiguous. I consider it more like a sign of a temporary workaround, or the last resort solution to plug-in some non-standard implementation.

amr-sc commented:
> > > > > any reason this can't be done with weak functions?
> 
> > > Is there a particular use for weak functions in this case?
> 
> > Hi @cfriedt , I think the _weak functions could be useful to override the default arch_ level implementations (which follow the Spec) in case some vendors doesn't conform to the Spec, or the Spec is ambiguous.
> 
> Sure - it's just that weak symbols are not part of ISO C or ISO C++ and are very much ELF specific.
> 
> We can achieve the same thing with a well-defined interface and having a Kconfig option to choose between the default (arch) and overridden (platform) solution.

Oh, I've got your point.
The explicit policy with default arch_ and overridden platform_ API is for sure better than the _weak approach.
Thanks! 

P.S.: looks like this may provide a guideline for #65824 as well

aescolar commented on doc/hardware/porting/custom.rst:
```diff
@@ -0,0 +1,16 @@
+When the provided support in Zephyr does not cover certain hardware platforms,
+some interfaces can be implemented directly by the platform or a custom
+architecture interface can be used and resulting in more flexibility and
+hardware support of differebt variants of the same hardware.
```
typo differebt

aescolar commented on doc/hardware/porting/custom.rst:
```diff
@@ -0,0 +1,16 @@
+
+When the provided support in Zephyr does not cover certain hardware platforms,
+some interfaces can be implemented directly by the platform or a custom
+architecture interface can be used and resulting in more flexibility and
```
".. can be used and resulting .."
This whole sentence is a bit too tough to read.

nashif commented:
> the more I look into things and try to fix existing inconsistencies and interfaces, I realize how messy interrupt implementation is and how many ways exist to enable interrupt controllers that we need to fix ASAP, so for example, I do not think we need to support platform specific irq handling as this PR does and instead allow for this to be done using dedicated interrupt drivers, the same way we do this already on many architectures.

just seen this #66505

ycsin commented:
@nashif any update for this PR?
